gnmi_subscribe/gnmi_subscribe.go

```diff
@@ -0,0 +1,192 @@
+	flag.Parse()
+
+	opts := credentials.ClientCredentials(*targetName)
+	conn, err := grpc.Dial(*targetAddr, opts...)
+	if err != nil {
+		log.Fatalf("Dialing to %s failed: %v", *targetAddr, err)
+	}
+	defer conn.Close()
+
+	client := pb.NewGNMIClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), *connectionTimeout*time.Second)
+	defer cancel()
+
+	subscribeClient, err = client.Subscribe(ctx)
+	if err != nil {
+		log.Fatalf("Error creating GNMI_SubscribeClient: %v", err)
+	}
+
+	encoding, ok := pb.Encoding_value[*encodingFormat]
+	if !ok {
+		var encodingList []string
+		for _, name := range pb.Encoding_name {
+			encodingList = append(encodingList, name)
+		}
+		log.Exitf("Supported encodings: %s", strings.Join(encodingList, ", "))
+	}
+
+	var subscriptionListMode gnmi.SubscriptionList_Mode
+	switch {
+	case *subscriptionPoll && *subscriptionOnce:
+		flag.Usage()
+		log.Exitf("Only one of -once and -poll can be set")
+	case *subscriptionOnce:
+		subscriptionListMode = pb.SubscriptionList_ONCE
+	case *subscriptionPoll:
+		subscriptionListMode = pb.SubscriptionList_POLL
+	default:
+		subscriptionListMode = pb.SubscriptionList_STREAM
+	}
+
+	var pbPathList []*pb.Path
+	for _, xPath := range xPathFlags {
+		pbPath, err := xpath.ToGNMIPath(xPath)
+		if err != nil {
+			log.Exitf("error in parsing xpath %q to gnmi path", xPath)
+		}
+		pbPathList = append(pbPathList, pbPath)
+	}
+	for _, textPbPath := range pbPathFlags {
+		var pbPath pb.Path
+		if err := proto.UnmarshalText(textPbPath, &pbPath); err != nil {
+			log.Exitf("error in unmarshaling %q to gnmi Path", textPbPath)
+		}
+		pbPathList = append(pbPathList, &pbPath)
+	}
+	subscriptions, err := assembleSubscriptions(pbPathList)
+	if err != nil {
+		log.Exitf("Error assembling subscriptions: %v", err)
+	}
+
+	request := &pb.SubscribeRequest{
+		Request: &pb.SubscribeRequest_Subscribe{
+			Subscribe: &pb.SubscriptionList{
+				Encoding:     pb.Encoding(encoding),
+				Mode:         subscriptionListMode,
+				Subscription: subscriptions,
+			},
+		},
+	}
+	utils.PrintProto(request)
+	if err := subscribeClient.Send(request); err != nil {
+		log.Exitf("Failed to send request: %v", err)
+	}
+	switch subscriptionListMode {
+	case pb.SubscriptionList_STREAM:
+		stream()
+	case pb.SubscriptionList_POLL:
+		poll()
+	case pb.SubscriptionList_ONCE:
+		once()
+	}
+}
+
+func stream() {
+
+}
+
+func poll() {
+
+}
+
+func once() {
+
+}
+
+func assembleSubscriptions(paths []*pb.Path) ([]*pb.Subscription, error) {
+	var subscriptions []*pb.Subscription
+	var subscriptionMode gnmi.SubscriptionMode
+	var heartbeat uint64
+	switch {
+	case *streamOnChange && *sampleInterval != 0:
+		return nil, errors.New("Only one of -stream_on_change and -sample_interval can be set")
+	case *streamOnChange:
+		subscriptionMode = pb.SubscriptionMode_ON_CHANGE
+	case *sampleInterval != 0:
+		subscriptionMode = pb.SubscriptionMode_SAMPLE
+		if *heartbeatInterval == 0 {
+			heartbeat = *sampleInterval * 10
```

Comment:
After thinking about it, I take it back, lets not enforce a default. Use it only if set.

Reply:
Sounds good!